src/middlewared/middlewared/main.py

```diff
@@ -1106,14 +1116,37 @@ def register_hook(self, name, method, *, inline=False, order=0, raise_error=Fals
 
         self.__hooks[name].append({
             'method': method,
+            'blockable': blockable,
             'inline': inline,
             'order': order,
             'raise_error': raise_error,
             'sync': sync,
         })
         self.__hooks[name] = sorted(self.__hooks[name], key=lambda hook: hook['order'])
 
+    @contextlib.contextmanager
+    def block_hooks(self, *names):
+        for name in names:
+            if not self.__hooks[name]:
```

Comment:
Will this not raise a `KeyError`?

Reply:
Also, it's kind of harsh to raise a `RuntimeError` here, right? Why not just log a warning so that middlewared service at least keeps running?

Reply:
> This will raise a `KeyError`, right?

neat, `collections.defaultdict` doesn't raise a `KeyError` so I've learned something new today.